bindata/network/additional-networks/crd/001-crd.yaml

```diff
@@ -16,7 +16,7 @@ spec:
   - name: v1
     served: true
     storage: true
-  preserveUnknownFields: true
+  preserveUnknownFields: false
```

Comment:
why does this affect `oc explain`? what is the output with and without this patch?

Reply:
without patch:

```
[tohayash@tohayash-lab tmp]$ oc version
Client Version: 4.5.4
Server Version: 4.6.0-0.ci-2020-08-05-165313
Kubernetes Version: v1.19.0-rc.2.442+6c41bb12ed9820-dirty
[tohayash@tohayash-lab tmp]$ oc get crd/network-attachment-definitions.k8s.cni.cncf.io -o yaml
apiVersion: apiextensions.k8s.io/v1
kind: CustomResourceDefinition
metadata:
  creationTimestamp: "2020-08-06T01:43:48Z"
  generation: 1
  managedFields:
    (snip)
  name: network-attachment-definitions.k8s.cni.cncf.io
  ownerReferences:
  - apiVersion: operator.openshift.io/v1
    blockOwnerDeletion: true
    controller: true
    kind: Network
    name: cluster
    uid: 9204b9e8-96a9-4bd5-ac29-acc12bf0f081
  resourceVersion: "42018"
  selfLink: /apis/apiextensions.k8s.io/v1/customresourcedefinitions/network-attachment-definitions.k8s.cni.cncf.io
  uid: 7e46f2da-1fd0-45f9-be5e-b60342b49e24
spec:
  conversion:
    strategy: None
  group: k8s.cni.cncf.io
  names:
    kind: NetworkAttachmentDefinition
    listKind: NetworkAttachmentDefinitionList
    plural: network-attachment-definitions
    shortNames:
    - net-attach-def
    singular: network-attachment-definition
  preserveUnknownFields: true
  scope: Namespaced
  versions:
  (snip)
[tohayash@tohayash-lab tmp]$ oc explain net-attach-def
KIND:     NetworkAttachmentDefinition
VERSION:  k8s.cni.cncf.io/v1

DESCRIPTION:
     <empty>
[tohayash@tohayash-lab tmp]$
```

Reply:
With patch:
```
[tohayash@tohayash-srv test200806a]$ oc version
Client Version: 4.5.0-0.ci-2020-07-29-002023
Server Version: 4.6.0-0.ci-2020-08-05-165313
Kubernetes Version: v1.19.0-rc.2.442+6c41bb12ed9820-dirty
[tohayash@tohayash-srv test200806a]$ oc get crd/network-attachment-definitions.k8s.cni.cncf.io -o yaml
apiVersion: apiextensions.k8s.io/v1
kind: CustomResourceDefinition
metadata:
  creationTimestamp: "2020-08-06T03:41:49Z"
  generation: 2
  managedFields:
  (snip)
  name: network-attachment-definitions.k8s.cni.cncf.io
  (snip)
spec:
  conversion:
    strategy: None
  group: k8s.cni.cncf.io
  names:
    kind: NetworkAttachmentDefinition
    listKind: NetworkAttachmentDefinitionList
    plural: network-attachment-definitions
    shortNames:
    - net-attach-def
    singular: network-attachment-definition
  scope: Namespaced
  versions:
  (snip)
[tohayash@tohayash-srv test200806a]$ oc explain net-attach-def
KIND:     NetworkAttachmentDefinition
VERSION:  k8s.cni.cncf.io/v1

DESCRIPTION:
     NetworkAttachmentDefinition is a CRD schema specified by the Network
     Plumbing Working Group to express the intent for attaching pods to one or
     more logical or physical networks. More information available at:
     https://github.com/k8snetworkplumbingwg/multi-net-spec

FIELDS:
   apiVersion   <string>
     APIVersion defines the versioned schema of this representation of an
     object. Servers should convert recognized schemas to the latest internal
     value, and may reject unrecognized values. More info:
     https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#resources

   kind <string>
     Kind is a string value representing the REST resource this object
     represents. Servers may infer this from the endpoint the client submits
     requests to. Cannot be updated. In CamelCase. More info:
     https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#types-kinds

   metadata     <Object>
     Standard object's metadata. More info:
     https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata

   spec <Object>
     NetworkAttachmentDefinition spec defines the desired state of a network
     attachment
```

Reply:
I am just guessing that the following code suppresses explain command output in case of `"x-kubernetes-preserve-unknown-fields: true` because explain command cannot explain unknown fields.

https://github.com/kubernetes/apiextensions-apiserver/blob/master/pkg/controller/openapi/builder/builder.go#L405-L409

Reply:
The latest k8s prefers to use `x-kubernetes-preserve-unknown-fields` instead of `preserveUnknownFields`, however I use `preserveUnknownFields` because of backporting. Once backports is done, I will update to `x-kubernetes-preserve-unknown-fields`  in master branch.

Reply:
Regarding why it's necessary to set the preserve unknown fields to false: https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#publish-validation-schema-in-openapi-v2